### PR TITLE
virttest.qemu_vm: fix use disconnect session in reboot

### PR DESCRIPTION
**Pull request by xutian**
disconnect session object used caused server-runing-lock
file not found error.

Signed-off-by: Xu Tian <xutian@redhat.com>

### CODE REVIEW

**Comment by xutian** [comment]
ID: 1342460

**Comment by xutian** [comment]
@mazhang  Please try this patch, I think this patch will fix issue your meet in testing.

Thanks,
Xu

**Comment by mazhang** [comment]
Just test with this patch, the problem has been fixed.

File: virttest/qemu_vm.py [hunk 1]
```diff
@@ -3790,62 +3790,74 @@ def reboot(self, session=None, method="shell", nic_index=0,
...
+            if not utils_misc.wait_for(lambda: __go_down(
+                    session, nic_index, serial), timeout=timeout):
+                if filter(lambda m: m.get_event('RESET'), qmp_monitors):
+                    logging.debug("QMP monitor recived 'RESET' event.")
+                raise virt_vm.VMRebootError("Guest refuses to go down")
+            else:
+                logging.info("VM go down, wait for boot up")
+
+        def shell_reboot(session, nic_index, serial, timeout):
+            """ Wait for VM go down """
```
Comment (CongLi) [inline virttest/qemu_vm.py]: I think the doc string is for wait_for_go_down().

File: virttest/qemu_vm.py [hunk 2]
```diff
@@ -4225,3 +4237,21 @@ def balloon(self, size):
...
+        qmp_monitors = [m for m in self.monitors if m.protocol == "qmp"]
+        for m in qmp_monitors:
+            m.clear_events()
+        self.monitor.cmd("system_reset")
+        if not qmp_monitors:
+            return True
+        end_time = time.time() + 5
+        while time.time() < end_time:
+            for m in qmp_monitors:
+                if m.get_event("RESET"):
```
Comment (CongLi) [inline virttest/qemu_vm.py]: It's better to implement event check in class QMPMonitor.

File: virttest/qemu_vm.py [hunk 3]
```diff
@@ -3790,62 +3790,74 @@ def reboot(self, session=None, method="shell", nic_index=0,
...
+                    patterns = [r".*[Rr]ebooting.*", r".*[Rr]estarting system.*",
+                                r".*[Mm]achine restart.*", r".*Linux version.*"]
+                    if session.read_until_any_line_matches(
+                            patterns, timeout=timeout):
+                        session.close()
+                        return True
+                else:
+                    session = __get_session(nic_index, serial)
+                    session.close()
+                    return False
```
Comment (CongLi) [inline virttest/qemu_vm.py]: I'm confused about why return False here, could you help explain it?

Comment (xutian) [inline virttest/qemu_vm.py]: if session is not a serial session and it is still alive, that mean guest still running not begin to reboot,  so go_down should be False.

**Comment by xutian** [comment]
@CongLi  updated, please review it. thanks

File: virttest/qemu_vm.py [hunk 4]
```diff
@@ -3790,62 +3790,86 @@ def reboot(self, session=None, method="shell", nic_index=0,
...
+        def wait_for_go_down(session, nic_index, serial, timeout):
+            """ Wait for VM go down
+            :param session: VM session object
+            :param nic_index: which nic what to login VM
+            :param serial: login VM via serial console or not
+            :param timeout: timeout wait for VM go down
+            """
+            if not utils_misc.wait_for(
+                    lambda: __go_down(session, nic_index, serial),
+                    timeout=timeout):
```
Comment (CongLi) [inline virttest/qemu_vm.py]: No need to pass timeout here.
It is not used in __go_down() .

File: virttest/qemu_vm.py [hunk 5]
```diff
@@ -3790,62 +3790,86 @@ def reboot(self, session=None, method="shell", nic_index=0,
...
-            login_timeout = timeout - 1
-            for m in qmp_monitors:
-                if m.get_event("RESET"):
-                    logging.info("RESET QMP event received")
-                else:
-                    raise virt_vm.VMRebootError("RESET QMP event not received "
-                                                "after system_reset "
-                                                "(monitor '%s')" % m.name)
+            go_down_timeout = timeout / 3
+            self.system_reset(go_down_timeout)
```
Comment (CongLi) [inline virttest/qemu_vm.py]: For system_reset(), there is no param.

File: virttest/qemu_vm.py (hunk 4 above)
Comment (xutian) [inline virttest/qemu_vm.py]: timeout is arg for wait_for function, not for __go_down.

**Comment by CongLi** [comment]
LGTM,
Acked-by: Cong Li <coli@redhat.com>

**Comment by XueqiangWei** [comment]
LGTM,
Acked-by: XueqiangWei <xuwei@redhat.com>

**Comment by xutian** [comment]
thanks @CongLi  @XueqiangWei

**Pull request merged**